File: backend/src/main/java/com/workflow/engine/page/PageValidator.java
package com.workflow.engine.page;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.workflow.api.dto.DataSourceMetadata;
import com.workflow.common.exception.BusinessException;
import com.workflow.engine.datasource.DataSourceDefinitionService;
import com.workflow.engine.datasource.entity.DataSourceDefinition;
import com.workflow.engine.form.column.ColumnConfig;
import com.workflow.engine.form.entity.FormDefinition;
import com.workflow.engine.form.repository.FormDefinitionRepository;
import com.workflow.engine.page.entity.PageDefinition;
import com.workflow.engine.tenant.TenantProvider;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * 页面发布校验器。
 * 校验 VIEW 页面发布前的绑定正确性：
 * - 绑定表单存在、已发布且为 BUSINESS 类型
 * - searchFields/columns 引用的列存在于绑定表单 column_config
 * - searchFields/columns 不得引用隐藏列；searchFields 不得引用 JSON/TEXT 列
 * PAGE 自定义页面（阶段二）：
 * - dataSources 条目 id 唯一、refId 指向存在且 ENABLED 的全局数据源
 * - rule 中数据组件 dataSourceId 命中 dataSources[].id
 * - actions set-filter 字段命中目标数据源 searchFields 白名单
 */
@Component
public class PageValidator {

    private final FormDefinitionRepository formDefRepository;
    private final ObjectMapper objectMapper;
    private final TenantProvider tenantProvider;
    private final DataSourceDefinitionService dsService;

    /** 不可作为查询条件的列类型（大字段） */
    private static final Set<String> NON_FILTERABLE_TYPES = Set.of("JSON", "TEXT", "LONGTEXT");

    /** 可声明数据源绑定的数据组件类型 */
    private static final Set<String> DATA_COMPONENT_TYPES = Set.of("page-table", "page-tree");

    public PageValidator(FormDefinitionRepository formDefRepository,
                         ObjectMapper objectMapper,
                         TenantProvider tenantProvider,
                         DataSourceDefinitionService dsService) {
        this.formDefRepository = formDefRepository;
        this.objectMapper = objectMapper;
        this.tenantProvider = tenantProvider;
        this.dsService = dsService;
    }

    /**
     * 校验页面可发布。
     *
     * <p>VIEW 绑定协议：dataSourceId（新协议，经 SPI metadata 取列）优先；
     * dataSourceId 为空且 formKey 非空（迁移被跳过的页）走遗留校验逻辑兜底。
     *
     * @param page 待发布页面（type=VIEW 或 PAGE）
     * @throws BusinessException 各类校验失败（错误码 400）
     */
    public void validateForPublish(PageDefinition page) {
        if ("PAGE".equals(page.getType())) {
            validateForPublishPage(page);
            return;
        }

        String tenantId = tenantProvider.getTenantId();
        String dataSourceId = page.getDataSourceId();
        String formKey = page.getFormKey();
        boolean hasDataSource = dataSourceId != null && !dataSourceId.isBlank();
        boolean hasFormKey = formKey != null && !formKey.isBlank();

        // 1. 绑定来源校验：数据源与遗留 formKey 至少其一
        if (!hasDataSource && !hasFormKey) {
            throw new BusinessException(400, "请选择数据源");
        }

        // 2. 取列并构建合法列 / 隐藏列 / 不可筛选列集合
        List<ColumnConfig> columns;
        if (hasDataSource) {
            // 新协议：数据源 metadata 取列（定义不存在/未启用 → dsService.metadata 抛 400/404）
            DataSourceMetadata meta = dsService.metadata(dataSourceId);
            columns = meta.getColumns() == null ? List.of() : meta.getColumns();
        } else {
            // 兼容兜底：formKey 遗留协议（绑定表单存在、已发布、BUSINESS 类型）
            Optional<FormDefinition> boundFormOpt = formDefRepository
                    .findFirstByTenantIdAndKeyAndStatusOrderByVersionDesc(tenantId, formKey, "PUBLISHED");
            if (boundFormOpt.isEmpty()) {
                throw new BusinessException(400, "绑定表单不存在或未发布: " + formKey);
            }
            FormDefinition boundForm = boundFormOpt.get();
            if (!"BUSINESS".equals(boundForm.getType())) {
                throw new BusinessException(400, "绑定表单 " + formKey + " 不是业务表单");
            }
            columns = parseColumnConfig(boundForm.getColumnConfig());
        }

        Set<String> validKeys = new HashSet<>();
        Set<String> hiddenKeys = new HashSet<>();
        Set<String> nonFilterableKeys = new HashSet<>();
        for (ColumnConfig c : columns) {
            validKeys.add(c.getKey());
            if (c.isHidden()) {
                hiddenKeys.add(c.getKey());
            }
            if (NON_FILTERABLE_TYPES.contains(c.getColumnType())) {
                nonFilterableKeys.add(c.getKey());
            }
        }

        // 3. 解析页面 schema
        JsonNode root = parseSchema(page.getSchema());
        JsonNode searchFields = root.path("searchFields");
        JsonNode pageColumns = root.path("columns");

        // 4. searchFields：引用列必须存在、非隐藏、非 JSON/TEXT
        if (searchFields.isArray()) {
            for (JsonNode field : searchFields) {
                String key = field.path("key").asText();
                if (!validKeys.contains(key)) {
                    throw new BusinessException(400, "查询字段引用列不存在: " + key);
                }
                if (hiddenKeys.contains(key)) {
                    throw new BusinessException(400, "查询字段不能引用隐藏列: " + key);
                }
                if (nonFilterableKeys.contains(key)) {
                    throw new BusinessException(400, "查询字段不能引用大字段列（JSON/TEXT）: " + key);
                }
            }
        }

        // 5. columns：引用列必须存在、非隐藏
        if (pageColumns.isArray()) {
            for (JsonNode column : pageColumns) {
                String key = column.path("key").asText();
                if (!validKeys.contains(key)) {
                    throw new BusinessException(400, "展示列引用列不存在: " + key);
                }
                if (hiddenKeys.contains(key)) {
                    throw new BusinessException(400, "展示列不能引用隐藏列: " + key);
                }
            }
        }
    }

    private void validateSchemaJson(String schema) {
        parseSchema(schema);
    }

    /**
     * 解析发布时编译视图所需的列（供 ViewCompiler 使用）。
     * 调用前提：validateForPublish 已通过。
     * dataSourceId 非空 → 数据源 metadata 列；否则 → 遗留 formKey 表单 column_config。
     *
     * @param page 待发布页面（type=VIEW）
     * @return 列定义列表
     * @throws BusinessException 数据源不可用或绑定表单不存在/未发布
     */
    public List<ColumnConfig> resolveBindColumns(PageDefinition page) {
        String dataSourceId = page.getDataSourceId();
        if (dataSourceId != null && !dataSourceId.isBlank()) {
            DataSourceMetadata meta = dsService.metadata(dataSourceId);
            return meta.getColumns() == null ? List.of() : meta.getColumns();
        }
        String tenantId = tenantProvider.getTenantId();
        if (page.getFormKey() == null || page.getFormKey().isBlank()) {
            throw new BusinessException(400, "视图必须绑定业务表单");
        }
        Optional<FormDefinition> boundFormOpt = formDefRepository
                .findFirstByTenantIdAndKeyAndStatusOrderByVersionDesc(tenantId, page.getFormKey(), "PUBLISHED");
        if (boundFormOpt.isEmpty()) {
            throw new BusinessException(400, "绑定表单不存在或未发布: " + page.getFormKey());
        }
        return parseColumnConfig(boundFormOpt.get().getColumnConfig());
    }

    private JsonNode parseSchema(String schema) throws BusinessException {
        try {
            JsonNode root = objectMapper.readTree(schema == null || schema.isBlank() ? "{}" : schema);
            // 兼容纯数组形式：{rule:[...]} 与裸数组
            if (root.isArray()) {
                return objectMapper.createObjectNode().set("rule", root);
            }
            return root;
        } catch (JsonProcessingException e) {
            throw new BusinessException(400, "页面 schema 解析失败");
        }
    }

    /**
     * 自定义页面（PAGE）发布校验：
     * 1. rule 可被 FormRenderer 解析（JSON 合法 + rule 数组）
     * 2. dataSources 条目 id 页面内唯一、refId 非空且指向存在且 ENABLED 的全局数据源
     * 3. rule 中数据组件（el-table/table）dataSourceId 命中 dataSources[].id
     * 4. actions set-filter 字段命中目标数据源 searchFields 白名单
     */
    private void validateForPublishPage(PageDefinition page) {
        String tenantId = tenantProvider.getTenantId();
        JsonNode root = parseSchema(page.getSchema());
        if (!root.isObject() || !root.path("rule").isArray()) {
            throw new BusinessException(400, "自定义页面 schema 必须为 {rule, option, dataSources, actions}");
        }

        // 已启用全局数据源映射（refId → DataSourceDefinition）
        Map<String, DataSourceDefinition> enabledDsMap = new HashMap<>();
        for (DataSourceDefinition ds : dsService.getEnabled()) {
            enabledDsMap.put(ds.getId(), ds);
        }

        // 2. dataSources 校验
        Map<String, JsonNode> dsById = new HashMap<>();
        JsonNode dataSources = root.path("dataSources");
        if (dataSources.isArray()) {
            for (JsonNode entry : dataSources) {
                String id = entry.path("id").asText();
                if (id.isBlank()) {
                    throw new BusinessException(400, "自定义页面 dataSources 条目 id 不能为空");
                }
                if (dsById.containsKey(id)) {
                    throw new BusinessException(400, "自定义页面 dataSources id 重复: " + id);
                }
                String refId = entry.path("refId").asText();
                if (refId.isBlank()) {
                    throw new BusinessException(400, "自定义页面 dataSources[" + id + "] refId 不能为空");
                }
                DataSourceDefinition ds = enabledDsMap.get(refId);
                if (ds == null) {
                    throw new BusinessException(400, "自定义页面引用的数据源不存在或未启用: " + refId);
                }
                // FORM 数据源：绑定表单须已发布；searchFields/columns 引用列存在
                if ("FORM".equals(ds.getType())) {
                    Optional<FormDefinition> formOpt = formDefRepository
                            .findFirstByTenantIdAndKeyAndStatusOrderByVersionDesc(
                                    tenantId, ds.getFormKey(), "PUBLISHED");
                    if (formOpt.isEmpty()) {
                        throw new BusinessException(400, "自定义页面引用的表单未发布: " + ds.getFormKey());
                    }
                    validatePageDsFields(entry, formOpt.get());
                }
                dsById.put(id, entry);
            }
        }

        // 3. rule 中数据组件 dataSourceId 命中
        JsonNode rule = root.path("rule");
        for (JsonNode node : rule) {
            String type = node.path("type").asText();
            if (!DATA_COMPONENT_TYPES.contains(type)) continue;
            String dsId = node.path("props").path("dataSourceId").asText();
            if (!dsId.isBlank() && !dsById.containsKey(dsId)) {
                throw new BusinessException(400, "数据组件 dataSourceId 未在 dataSources 声明: " + dsId);
            }
        }

        // 4. actions set-filter 字段白名单
        JsonNode actions = root.path("actions");
        if (actions.isArray()) {
            for (JsonNode action : actions) {
                JsonNode steps = action.path("steps");
                if (!steps.isArray()) continue;
                for (JsonNode step : steps) {
                    if (!"set-filter".equals(step.path("op").asText())) continue;
                    String target = step.path("target").asText();
                    String field = step.path("field").asText();
                    if (target.isBlank() || field.isBlank()) continue;
                    JsonNode dsEntry = dsById.get(target);
                    if (dsEntry == null) {
                        throw new BusinessException(400, "set-filter 目标数据源未声明: " + target);
                    }
                    JsonNode searchFields = dsEntry.path("searchFields");
                    if (searchFields.isArray() && searchFields.size() > 0) {
                        boolean declared = false;
                        for (JsonNode sf : searchFields) {
                            if (field.equals(sf.asText())) { declared = true; break; }
                        }
                        if (!declared) {
                            throw new BusinessException(400, "set-filter 字段未在数据源 searchFields 白名单: " + field);
                        }
                    }
                }
            }
        }
    }

    /** 校验 PAGE dataSources 条目的 searchFields/columns 引用列存在（FORM 绑定） */
    private void validatePageDsFields(JsonNode dsEntry, FormDefinition form) {
        List<ColumnConfig> columns = parseColumnConfig(form.getColumnConfig());
        Set<String> validKeys = new HashSet<>();
        for (ColumnConfig c : columns) {
            validKeys.add(c.getKey());
        }
        JsonNode searchFields = dsEntry.path("searchFields");
        if (searchFields.isArray()) {
            for (JsonNode sf : searchFields) {
                String key = sf.asText();
                if (!key.isBlank() && !validKeys.contains(key)) {
                    throw new BusinessException(400, "数据源 searchFields 引用列不存在: " + key);
                }
            }
        }
        JsonNode pageColumns = dsEntry.path("columns");
        if (pageColumns.isArray()) {
            for (JsonNode col : pageColumns) {
                String key = col.asText();
                if (!key.isBlank() && !validKeys.contains(key)) {
                    throw new BusinessException(400, "数据源 columns 引用列不存在: " + key);
                }
            }
        }
    }

    private List<ColumnConfig> parseColumnConfig(String columnConfig) throws BusinessException {
        if (columnConfig == null || columnConfig.isBlank()) {
            throw new BusinessException(400, "绑定表单未配置列映射（column_config）");
        }
        try {
            return objectMapper.readValue(columnConfig,
                    objectMapper.getTypeFactory().constructCollectionType(List.class, ColumnConfig.class));
        } catch (JsonProcessingException e) {
            throw new BusinessException(400, "绑定表单列映射配置非法");
        }
    }
}

File: backend-node/src/engine/page/page-validator.ts
import { Injectable } from '@nestjs/common'
import { BusinessException } from '../../common/exception/business-exception'
import type { ColumnConfig } from '../../common/domain/column-config'
import { getTenantId } from '../../framework/tenant/tenant-context'
import { parseBusinessColumnConfig } from '../form/column/column-config-parser'
import { FormDefinitionRepository } from '../form/repository/form-definition.repository'
import { DataSourceService } from '../datasource/service/data-source.service'
import type { PageDefinitionRow } from './repository/page-definition.repository'

/** 不可作为查询条件的列类型（大字段），对齐 Java `NON_FILTERABLE_TYPES`。 */
const NON_FILTERABLE_TYPES = new Set(['JSON', 'TEXT', 'LONGTEXT'])

/** 可声明数据源绑定的数据组件类型，对齐 Java `DATA_COMPONENT_TYPES`。 */
const DATA_COMPONENT_TYPES = new Set(['page-table', 'page-tree'])

/**
 * 页面发布校验器（对齐 Java `PageValidator`，331 行）。
 *
 * 两套互不相交的规则，按 `type` 分派：
 * - **VIEW**：绑定来源（dataSourceId 优先，formKey 兜底）→ 取列 → schema 里的
 *   `searchFields` / `columns` 逐个比对列白名单（存在、非隐藏、非大字段）；
 * - **PAGE**：`{rule, option, dataSources, actions}` 结构 → `dataSources[].refId`
 *   指向**存在且 ENABLED** 的全局数据源 → `rule` 里数据组件的 `dataSourceId` 已声明
 *   → `actions` 的 `set-filter` 字段命中该数据源的 `searchFields` 白名单。
 *
 * ⚠️ 校验项**逐条照抄**是本文件的全部价值：漏一条的后果是"非法 schema 被存进库"，
 *    而它要到很久以后（渲染或取数）才炸，那时已经很难定位来源。
 *    Java 侧 14 处 `BusinessException` 的消息文本在这里逐字复刻。
 *
 * ⚠️ `parseColumnConfig` 用**已有的** `parseBusinessColumnConfig`，不另写一份：
 *    它内部从 `newColumnConfig()` 起步补齐默认值（Jackson 语义）。
 *    Java 用的是 `objectMapper.readValue(..., ColumnConfig.class)` ——
 *    两者对**缺省字段**的处理必须一致，否则「表单未配置某列属性」会被判成非法。
 */
@Injectable()
export class PageValidator {
  constructor(
    private readonly formDefRepository: FormDefinitionRepository,
    private readonly dataSourceService: DataSourceService,
  ) {}

  /**
   * 校验页面可发布（对齐 Java `validateForPublish`）。
   *
   * `PAGE` 类型走另一条完全独立的规则集。
   */
  async validateForPublish(page: PageDefinitionRow): Promise<void> {
    if (page.type === 'PAGE') {
      await this.validateForPublishPage(page)
      return
    }

    const tenantId = getTenantId()
    const hasDataSource = page.data_source_id !== null && page.data_source_id.trim() !== ''
    const hasFormKey = page.form_key !== null && page.form_key.trim() !== ''

    // 1. 绑定来源校验：数据源与遗留 formKey 至少其一
    if (!hasDataSource && !hasFormKey) {
      throw new BusinessException(400, '请选择数据源')
    }

    // 2. 取列并构建合法列 / 隐藏列 / 不可筛选列集合
    const columns = await this.resolveColumns(
      tenantId,
      page.data_source_id,
      page.form_key,
      '绑定表单不存在或未发布: ',
    )

    const validKeys = new Set<string>()
    const hiddenKeys = new Set<string>()
    const nonFilterableKeys = new Set<string>()
    for (const column of columns) {
      const key = String(column.key)
      validKeys.add(key)
      if (column.hidden) hiddenKeys.add(key)
      if (NON_FILTERABLE_TYPES.has(String(column.columnType))) nonFilterableKeys.add(key)
    }

    // 3. 解析页面 schema
    const root = parseSchema(page.schema)
    const searchFields = pathOf(root, 'searchFields')
    const pageColumns = pathOf(root, 'columns')

    // 4. searchFields：引用列必须存在、非隐藏、非 JSON/TEXT
    if (Array.isArray(searchFields)) {
      for (const field of searchFields) {
        const key = textAt(field, 'key')
        if (!validKeys.has(key)) {
          throw new BusinessException(400, `查询字段引用列不存在: ${key}`)
        }
        if (hiddenKeys.has(key)) {
          throw new BusinessException(400, `查询字段不能引用隐藏列: ${key}`)
        }
        if (nonFilterableKeys.has(key)) {
          throw new BusinessException(400, `查询字段不能引用大字段列（JSON/TEXT）: ${key}`)
        }
      }
    }

    // 5. columns：引用列必须存在、非隐藏（自定义列 custom=true 跳过数据源字段校验）
    if (Array.isArray(pageColumns)) {
      for (const column of pageColumns) {
        const key = textAt(column, 'key')
        const isCustom = boolAt(column, 'custom')
        if (!isCustom && !validKeys.has(key)) {
          throw new BusinessException(400, `展示列引用列不存在: ${key}`)
        }
        if (hiddenKeys.has(key)) {
          throw new BusinessException(400, `展示列不能引用隐藏列: ${key}`)
        }
      }
    }
  }

  /**
   * 解析发布时编译视图所需的列（对齐 Java `resolveBindColumns`）。
   * 调用前提：`validateForPublish` 已通过。
   *
   * ⚠️ 仅供 VIEW 编译用；本批**没有**迁移 `ViewCompiler`（规格 U32），
   *    所以这个方法的调用方目前只有 `publish` 的显式拒绝分支。
   */
  async resolveBindColumns(page: PageDefinitionRow): Promise<ColumnConfig[]> {
    if (page.data_source_id !== null && page.data_source_id.trim() !== '') {
      const meta = await this.dataSourceService.metadata(page.data_source_id)
      return meta.columns ?? []
    }
    if (page.form_key === null || page.form_key.trim() === '') {
      throw new BusinessException(400, '视图必须绑定业务表单')
    }
    return this.resolveColumns(
      getTenantId(),
      null,
      page.form_key,
      '绑定表单不存在或未发布: ',
    )
  }

  /** 取列：数据源 metadata 优先，其次遗留 formKey 的 column_config。 */
  private async resolveColumns(
    tenantId: string,
    dataSourceId: string | null,
    formKey: string | null,
    missingFormMessage: string,
  ): Promise<ColumnConfig[]> {
    if (dataSourceId !== null && dataSourceId.trim() !== '') {
      // 新协议：数据源 metadata 取列（定义不存在/未启用 → dsService 自带 404/400）
      const meta = await this.dataSourceService.metadata(dataSourceId)
      return meta.columns ?? []
    }
    const boundForm = await this.formDefRepository.findLatestPublishedByKey(
      String(formKey),
      tenantId,
    )
    if (boundForm === null) {
      throw new BusinessException(400, `${missingFormMessage}${String(formKey)}`)
    }
    if (boundForm.type !== 'BUSINESS') {
      throw new BusinessException(400, `绑定表单 ${String(formKey)} 不是业务表单`)
    }
    return parseColumnConfig(boundForm.column_config)
  }

  /**
   * 自定义页面（PAGE）发布校验（对齐 Java `validateForPublishPage`）。
   *
   * 四步：schema 形状 → dataSources 条目（id 唯一、refId 非空且指向 ENABLED 数据源、
   * FORM 数据源还要求绑定表单已发布且 searchFields/columns 引用列存在）
   * → rule 中数据组件的 dataSourceId 命中 → actions 的 set-filter 字段白名单。
   */
  private async validateForPublishPage(page: PageDefinitionRow): Promise<void> {
    const tenantId = getTenantId()
    const root = parseSchema(page.schema)
    if (!Array.isArray(pathOf(root, 'rule'))) {
      throw new BusinessException(400, '自定义页面 schema 必须为 {rule, option, dataSources, actions}')
    }

    // 已启用全局数据源映射（refId → 数据源）
    const enabled = await this.dataSourceService.getEnabled()
    const enabledById = new Map(enabled.map((ds) => [ds.id, ds]))

    // 2. dataSources 校验
    const dsById = new Map<string, Record<string, unknown>>()
    const dataSources = pathOf(root, 'dataSources')
    if (Array.isArray(dataSources)) {
      for (const entry of dataSources) {
        const record = asRecord(entry)
        const id = textAt(record, 'id')
        if (id.trim() === '') {
          throw new BusinessException(400, '自定义页面 dataSources 条目 id 不能为空')
        }
        if (dsById.has(id)) {
          throw new BusinessException(400, `自定义页面 dataSources id 重复: ${id}`)
        }
        const refId = textAt(record, 'refId')
        if (refId.trim() === '') {
          throw new BusinessException(400, `自定义页面 dataSources[${id}] refId 不能为空`)
        }
        const ds = enabledById.get(refId)
        if (ds === undefined) {
          throw new BusinessException(400, `自定义页面引用的数据源不存在或未启用: ${refId}`)
        }
        // FORM 数据源：绑定表单须已发布；searchFields/columns 引用列存在
        if (ds.type === 'FORM') {
          const boundForm = await this.formDefRepository.findLatestPublishedByKey(
            String(ds.formKey),
            tenantId,
          )
          if (boundForm === null) {
            throw new BusinessException(400, `自定义页面引用的表单未发布: ${String(ds.formKey)}`)
          }
          validatePageDsFields(record, boundForm.column_config)
        }
        dsById.set(id, record)
      }
    }

    // 3. rule 中数据组件 dataSourceId 命中
    const rule = pathOf(root, 'rule')
    for (const node of Array.isArray(rule) ? rule : []) {
      const type = textAt(node, 'type')
      if (!DATA_COMPONENT_TYPES.has(type)) continue
      const dsId = textAt(pathOf(node, 'props'), 'dataSourceId')
      if (dsId.trim() !== '' && !dsById.has(dsId)) {
        throw new BusinessException(400, `数据组件 dataSourceId 未在 dataSources 声明: ${dsId}`)
      }
    }

    // 4. actions set-filter 字段白名单
    const actions = pathOf(root, 'actions')
    for (const action of Array.isArray(actions) ? actions : []) {
      const steps = pathOf(action, 'steps')
      for (const step of Array.isArray(steps) ? steps : []) {
        if (textAt(step, 'op') !== 'set-filter') continue
        const target = textAt(step, 'target')
        const field = textAt(step, 'field')
        if (target.trim() === '' || field.trim() === '') continue
        const entry = dsById.get(target)
        if (entry === undefined) {
          throw new BusinessException(400, `set-filter 目标数据源未声明: ${target}`)
        }
        const searchFields = pathOf(entry, 'searchFields')
        if (Array.isArray(searchFields) && searchFields.length > 0) {
          const declared = searchFields.some((sf) => field === String(sf))
          if (!declared) {
            throw new BusinessException(400, `set-filter 字段未在数据源 searchFields 白名单: ${field}`)
          }
        }
      }
    }
  }
}

/** 校验 PAGE dataSources 条目的 searchFields/columns 引用列存在（FORM 绑定）。 */
function validatePageDsFields(entry: Record<string, unknown>, columnConfig: string | null): void {
  const validKeys = new Set(parseColumnConfig(columnConfig).map((c) => String(c.key)))
  const searchFields = pathOf(entry, 'searchFields')
  for (const sf of Array.isArray(searchFields) ? searchFields : []) {
    const key = String(sf)
    if (key.trim() !== '' && !validKeys.has(key)) {
      throw new BusinessException(400, `数据源 searchFields 引用列不存在: ${key}`)
    }
  }
  const columns = pathOf(entry, 'columns')
  for (const col of Array.isArray(columns) ? columns : []) {
    const key = String(col)
    if (key.trim() !== '' && !validKeys.has(key)) {
      throw new BusinessException(400, `数据源 columns 引用列不存在: ${key}`)
    }
  }
}

/**
 * 解析 schema JSON（对齐 Java `parseSchema`）。
 *
 * ⚠️ 兼容**纯数组**形式：裸数组会被包成 `{rule: [...]}`。
 * ⚠️ 空 / 空白 schema 视为 `{}`（不是报错）；非法 JSON → 400「页面 schema 解析失败」。
 */
function parseSchema(schema: string | null): Record<string, unknown> {
  if (schema === null || schema.trim() === '') return {}
  let parsed: unknown
  try {
    parsed = JSON.parse(schema)
  } catch {
    throw new BusinessException(400, '页面 schema 解析失败')
  }
  if (Array.isArray(parsed)) return { rule: parsed }
  if (parsed === null || typeof parsed !== 'object') return {}
  return parsed as Record<string, unknown>
}

/**
 * 解析 column_config（空 → 400；非法 JSON → 400），复用既有的解析器补齐默认值。
 *
 * ⚠️ **已知偏差**：`parseBusinessColumnConfig` 比 Jackson 严格 —— 它对
 *    非法列名 / 保留列 / 未知列类型也会抛 `BusinessException`，而 Java 的
 *    `objectMapper.readValue(..., ColumnConfig.class)` 只关心**反序列化**，这类
 *    语义问题它根本不看（要么留给下游，要么永远不报）。这里把两者统一映射成
 *    Java 那条消息，代价是「非法列名」这类情形在 VIEW 发布时会变成
 *    「绑定表单列映射配置非法」而不是照常通过。
 *    该分支（VIEW + 只有 formKey 的遗留页）当前没有 golden 覆盖、且被 U32 挡在后面，
 *    属规格 §9 的开放项 —— **不要在没有 golden 的情况下擅自选择其中一种语义**。
 */
function parseColumnConfig(columnConfig: string | null): ColumnConfig[] {
  if (columnConfig === null || columnConfig.trim() === '') {
    throw new BusinessException(400, '绑定表单未配置列映射（column_config）')
  }
  try {
    return parseBusinessColumnConfig(columnConfig)
  } catch {
    throw new BusinessException(400, '绑定表单列映射配置非法')
  }
}

/** 取一个「可能是对象」的值的属性（对齐 Jackson 的 `path(...)`：非对象 → 缺失）。 */
function pathOf(node: unknown, field: string): unknown {
  if (node === null || typeof node !== 'object' || Array.isArray(node)) return undefined
  return (node as Record<string, unknown>)[field]
}

/** 对齐 Jackson 的 `path(x).asText()`：缺失/null → `""`。 */
function textAt(node: unknown, field: string): string {
  const value = pathOf(node, field)
  if (value === null || value === undefined) return ''
  return String(value)
}

/** 对齐 Jackson 的 `path(x).asBoolean(default)`：仅真正布尔为 true。 */
function boolAt(node: unknown, field: string): boolean {
  return pathOf(node, field) === true
}

/** 非对象 → 空对象（后续 `pathOf` 会返回 undefined，与 Jackson 语义一致）。 */
function asRecord(node: unknown): Record<string, unknown> {
  if (node === null || typeof node !== 'object' || Array.isArray(node)) return {}
  return node as Record<string, unknown>
}
